### PR TITLE
Spread relay load across all validator pigeons

### PR DESCRIPTION
**Pull request by MechanicalTyler**
# Related Github tickets

- https://github.com/VolumeFi/paloma/issues/299

# Background

Instead of picking a single pigeon to relay messages every 10 blocks, we are now having all pigeons relay some of the messages.  Each pigeon in the current valset gets a subset of the messages to relay.

This should help speed up relaying

#### Additionally, there is a fix for undesired behavior that was discovered.

Previously, if there was an error during relay, the current pigeon would
attest the message and then never touch it again.  This means that for
the message to get enough attestations, many pigeons had to attempt the
relay.  This is slow, inefficient, and no longer works with the new way
of dividing up messages.

Now, if a relayer hits an error message, it will send that error data
back to Paloma and other pigeons can simply attest to the error message.
This removes the need to attempt the relay by lots of pigeons to clear
out a bad message.

# Testing completed

- [x] wrote tests
- [x] tested on a local private testnet

### CODE REVIEW

File: relayer/message_relayer.go [hunk 1]
```diff
@@ -30,16 +28,33 @@ func (r *Relayer) RelayMessages(ctx context.Context, locker sync.Locker) error {
...
+	return 0, 0, chain.ErrNotValidator
+}
+
 func (r *Relayer) relayMessages(ctx context.Context, processors []chain.Processor) error {
 	if len(processors) == 0 {
 		return nil
 	}
 
-	ctx, cleanup, err := collision.GoStartLane(ctx, r.palomaClient, r.palomaClient.GetValidatorAddress())
+	// Get the offset for this validator so that we know which messages to relay
```
Comment (taariq) [inline relayer/message_relayer.go:53]: Could this risk that some validators NEVER get any messages to relay?

Comment (ToasterTheBrave) [reply relayer/message_relayer.go:53]: There shouldn't be any risk of that.  The msg IDs are always increasing, they never reset.  Each new msg will go to the next validator in the list.  The first validator won't get another message until the last validator has see none.

**Comment by ToasterTheBrave** [comment]
@taariq Your reflection is accurate.  This switches to 1 attempt by 1 pigeon.  It's a switch from "at least once execution" to "at most once execution".  If a pigeon fails to relay a message, then that message is not retried and we would instead want a new message.

The majority (actually, probably almost 100%) of our current errors are errors returned from the smart contract that won't succeed on retry.  They have nothing to do with the pigeon.  This solves that problem in a fairly simple and understandable way.  We can make it more complex and implement retries by different pigeons if you'd like, but I think it should be a follow-up feature after we take these speed improvements.

**Comment by taariq** [comment]
Ideally, we really want pigeons to select based on fee and then round-robin. So yes, this will be an optimization to the structure of the messaging marketplace. However, we must always assume that pigeons are not good nor rational actors in the system. Some pigeons will fail message delivery just so they can watch the world burn. We do not slash or jail for message delivery failure. However, we can even separate failures from pigeon vs. failures from the smart contract and treat them differently. Again, another optimization that we'll need to revisit, but we don't want to block a functional system. Okay. I'm good with smart contract failure as one-attempt, at most.

**Comment by ToasterTheBrave** [comment]
@taariq This has had another commit, and therefore could use another review before I merge

**Review by taariq — APPROVED** [review]
To reflect: We no longer have all pigeons attempting failed messages. We have at most one attempt, and then we send the error details back to Paloma with the proof of the error. Querying messages in the queue now returns all messages that are not yet attested. We also have improved some error logging here. Did I miss anything?

**Comment by ToasterTheBrave** [comment]
> To reflect: We no longer have all pigeons attempting failed messages. We have at most one attempt, and then we send the error details back to Paloma with the proof of the error. Querying messages in the queue now returns all messages that are not yet attested. We also have improved some error logging here. Did I miss anything?

Didn't miss anything.  You're spot on!

**Pull request merged**